File: apps/web/lib/portal-features.ts
// Which functions a client's portal offers. The agency switches each one per client
// (the toggles on the client's tabs); the portal shows only what is on, and the API
// refuses the rest. The catalog below mirrors apps/api/app/portal_features.py: a test
// compares the two lists, so change both together.

import { CHANNEL_TYPES, type ChannelType } from "@/lib/routes";

export const PORTAL_FEATURES = [
  { key: "inbox", default: true },
  { key: "contacts", default: true },
  { key: "pipeline", default: true },
  { key: "calendar", default: true },
  { key: "reports", default: true },
  { key: "teams", default: true },
  { key: "tags", default: true },
  { key: "templates", default: true },
  { key: "canned", default: true },
  { key: "agents", default: false },
  { key: "api", default: false },
  { key: "channels.whatsapp", default: false },
  { key: "channels.whatsapp_cloud", default: false },
  { key: "channels.instagram", default: false },
  { key: "channels.messenger", default: false },
  { key: "channels.webchat", default: false },
] as const;

export type PortalFeature = (typeof PORTAL_FEATURES)[number]["key"];

/** Functions whose portal screen is not built yet: the switch is stored, nothing shows. */
export const FEATURES_WITHOUT_SCREEN: readonly PortalFeature[] = ["api"];

export const CHANNEL_FEATURES: readonly PortalFeature[] = ["channels.whatsapp", "channels.whatsapp_cloud", "channels.instagram", "channels.messenger", "channels.webchat"];

/** The function behind each channel type's portal page (the type is the path segment of lib/routes.ts). */
export const FEATURE_OF_CHANNEL_TYPE: Record<ChannelType, PortalFeature> = {
  whatsapp: "channels.whatsapp",
  "whatsapp-cloud": "channels.whatsapp_cloud",
  instagram: "channels.instagram",
  messenger: "channels.messenger",
  webchat: "channels.webchat",
};

/** The switches shown on each tab of the client page; the Portal tab shows all of them. */
export const FEATURES_BY_CLIENT_TAB: Record<string, readonly PortalFeature[]> = {
  inbox: ["inbox"],
  agents: ["agents"],
  channels: CHANNEL_FEATURES,
  teams: ["teams"],
  tags: ["tags"],
  templates: ["templates"],
  calendar: ["calendar"],
  pipeline: ["pipeline"],
  api: ["api"],
  portal: PORTAL_FEATURES.map((entry) => entry.key),
};

/** A role permission is only as good as the function behind it. */
export const FEATURE_OF_PERMISSION: Record<string, PortalFeature> = {
  "contacts.manage": "contacts",
  "pipeline.manage": "pipeline",
  "calendar.manage": "calendar",
  "reports.view": "reports",
  "teams.manage": "teams",
  "tags.manage": "tags",
  "templates.manage": "templates",
  "canned.manage": "canned",
  "agents.manage": "agents",
};

/** Permissions that hold while ANY of several functions is on (channel management needs at least one channel type). */
const ANY_FEATURE_OF_PERMISSION: Record<string, readonly PortalFeature[]> = {
  "channels.manage": CHANNEL_FEATURES,
};

/** Whether the function behind a permission is on; a permission with no function behind it always is. */
export function permissionFeatureOn(permission: string, enabled: (key: PortalFeature) => boolean): boolean {
  const single = FEATURE_OF_PERMISSION[permission];
  if (single) return enabled(single);
  const any = ANY_FEATURE_OF_PERMISSION[permission];
  return any ? any.some(enabled) : true;
}

/** The channel types switched on for the portal, in the order the overview shows them. */
export function enabledChannelTypes(enabled: (key: PortalFeature) => boolean): ChannelType[] {
  return CHANNEL_TYPES.filter((type) => enabled(FEATURE_OF_CHANNEL_TYPE[type]));
}

const DEFAULTS = new Map<string, boolean>(PORTAL_FEATURES.map((entry) => [entry.key, entry.default]));

/** The client's full set, with any missing key at its default. */
export function normalizeFeatures(stored?: Record<string, boolean> | null): Record<PortalFeature, boolean> {
  const result = {} as Record<PortalFeature, boolean>;
  for (const entry of PORTAL_FEATURES) result[entry.key] = typeof stored?.[entry.key] === "boolean" ? Boolean(stored[entry.key]) : entry.default;
  return result;
}

/** Whether a portal session has a function. A session from an older server carries no list: defaults apply. */
export function hasFeature(enabled: readonly string[] | undefined, key: PortalFeature): boolean {
  return enabled === undefined ? Boolean(DEFAULTS.get(key)) : enabled.includes(key);
}

File: apps/api/app/portal_features.py
"""Which functions exist inside a client's portal.

The agency decides this client by client. The catalog below is the single
source of truth for the keys and their defaults; the web app keeps a typed
mirror in ``apps/web/lib/portal-features.ts`` and a test fails if the two
drift. A client stores only what the agency chose; anything missing reads as
the default, so adding a key to the catalog later never needs a data
migration, and a key that leaves the catalog is ignored where it is stored.

The first nine switches govern screens the portal already has, so they default
to on: every existing client keeps seeing exactly what it saw before. The
rest belong to screens the agency opens up per client and default to off.

Features only govern the portal (``/portal/{slug}/...``). The agency panel and
the public API are never affected by them.
"""

from typing import Any

from fastapi import HTTPException

FEATURE_DISABLED_DETAIL = "This feature is not enabled for this portal"

# (key, default), in display order.
CATALOG: tuple[tuple[str, bool], ...] = (
    ("inbox", True),
    ("contacts", True),
    ("pipeline", True),
    ("calendar", True),
    ("reports", True),
    ("teams", True),
    ("tags", True),
    ("templates", True),
    ("canned", True),
    ("agents", False),
    ("api", False),
    ("channels.whatsapp", False),
    ("channels.whatsapp_cloud", False),
    ("channels.instagram", False),
    ("channels.messenger", False),
    ("channels.webchat", False),
    ("details", False),
    ("professionals", False),
    ("services", False),
    ("appointments", False),
)

KEYS: tuple[str, ...] = tuple(key for key, _ in CATALOG)
DEFAULTS: dict[str, bool] = dict(CATALOG)


def defaults() -> dict[str, bool]:
    """A fresh copy of the defaults, safe to store or mutate."""
    return dict(DEFAULTS)


def normalize(stored: dict | None) -> dict[str, bool]:
    """The full set of switches: defaults, overridden by what is stored.

    Unknown keys are dropped and values are coerced to bool, so whatever an
    older or hand-edited row holds can never leak into a response.
    """
    result = defaults()
    if isinstance(stored, dict):
        for key in KEYS:
            if key in stored:
                result[key] = bool(stored[key])
    return result


def enabled_keys(client: Any) -> list[str]:
    """The keys switched on for ``client``, in catalog order."""
    features = normalize(getattr(client, "portal_features", None))
    return [key for key in KEYS if features[key]]


def is_enabled(client: Any, key: str) -> bool:
    return normalize(getattr(client, "portal_features", None)).get(key, False)


def validate_patch(patch: dict) -> dict[str, bool]:
    """Check an agency's partial update: known keys, real booleans."""
    if not isinstance(patch, dict):
        raise HTTPException(status_code=422, detail="portal_features must be an object of switches")
    unknown = sorted(str(key) for key in patch if key not in DEFAULTS)
    if unknown:
        raise HTTPException(status_code=422, detail=f"Unknown portal feature: {', '.join(unknown)}")
    for key, value in patch.items():
        if not isinstance(value, bool):
            raise HTTPException(status_code=422, detail=f"Portal feature '{key}' must be true or false")
    return {key: value for key, value in patch.items()}


def merged(stored: dict | None, patch: dict) -> dict[str, bool]:
    """The stored switches (normalized) with ``patch`` applied on top."""
    return {**normalize(stored), **validate_patch(patch)}


def ensure_enabled(client: Any, key: str) -> None:
    if not is_enabled(client, key):
        raise HTTPException(status_code=403, detail=FEATURE_DISABLED_DETAIL)
